### PR TITLE
add --forbid-only and --forbid-pending options

### LINKED ISSUE #1818: add --strict flag
I'd like a `--strict` CLI flag that will cause my tests to exit with non-zero status if
- any test is marked as only
- any test is marked as skipped
- any test is marked as pending

### PR DESCRIPTION
**Pull request by charlierudolph**
resolves #1818

### CODE REVIEW

**Comment by Munter** [comment]
Looks good to me. I haven't been to involved with this issue, so I'd like a second opinion. Ping @hollomancer @ScottFreeCode @dasilvacontin

**Comment by ScottFreeCode** [comment]
I like the idea in general, and this code (with tests!) looks pretty good. Only thing I wonder about is whether there should be a test case with `this.skip` in addition to `it.skip` and `it("name" /* and no test function */)`. However, in the discussion in #1818 and #1935 there was some concern that "production", while less liable to confusion than "strict", doesn't necessarily correlate well to the likely use cases, and there was disagreement as to whether skipped/pending should be included.

I'd like to propose instead having two separate flags named explicitly for what they do: `--forbid-only` and `--forbid-skip`/`--forbid-pending`.

**Comment by charlierudolph** [comment]
Two options are fine by me. I like `--forbid-only` and `--forbid-pending` (since skipped tests are marked as pending in that stats)

**Comment by charlierudolph** [comment]
In the next major release, I also think it would be a good idea to make forbid only and forbid pending the default and have the options be `--allow-only` and `--allow-pending`

**Comment by charlierudolph** [comment]
@ScottFreeCode this is now updated with the new options

**Comment by dasilvacontin** [comment]
> In the next major release, I also think it would be a good idea to make forbid only and forbid pending the default and have the options be --allow-only and --allow-pending

Added only when ENV=production or CI=true, or?

**Comment by charlierudolph** [comment]
@ScottFreeCode any idea when this is going to get merged?

**Comment by ScottFreeCode** [comment]
Did we end up having any concerns from @hollomancer or @dasilvacontin? I see @dasilvacontin had a question about future, backwards-incompatible default version of the behavior, but I presume that need not block adding the current opt-in version.

**Comment by dasilvacontin** [comment]
@ScottFreeCode With "question about future, backwards-incompatible default version of the behavior" are you referring to the last bit at https://github.com/mochajs/mocha/pull/1935#issuecomment-209350006, or to which comment are you referring to? Cheers!

**Comment by dasilvacontin** [comment]
Happy to see this moving forward btw! :)

**Comment by ScottFreeCode** [comment]
@dasilvacontin I was referring to this one: https://github.com/mochajs/mocha/pull/2696#issuecomment-277636371

**Comment by dasilvacontin** [comment]
@ScottFreeCode I see. Yeah, it doesn't make a difference for this PR – we can discuss it later. What I don't like about using ENV vars is that it doesn't stop you from committing the `.only`s. You'll notice them because of CI, but you are committing them nevertheless.

File: lib/runner.js [hunk 1]
```diff
@@ -821,9 +821,16 @@ Runner.prototype.run = function (fn) {
...
+    if (self.forbidOnly && self.hasOnly) {
+      failures += self.stats.tests;
+    }
+    if (self.forbidPending) {
+      failures += self.stats.pending;
+    }
     debug('end');
     process.removeListener('uncaughtException', uncaught);
-    fn(self.failures);
+    fn(failures);
```
Comment (charlierudolph) [inline lib/runner.js]: fixed

File: test/integration/fixtures/options/forbid-only/passed.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+describe('forbid only - all test pass', function () {
```
Comment (charlierudolph) [inline test/integration/fixtures/options/forbid-only/passed.js]: fixed

**Comment by charlierudolph** [comment]
@dasilvacontin comments addressed

File: bin/_mocha [hunk 3]
```diff
@@ -112,6 +112,8 @@ program
   .option('--watch-extensions <ext>,...', 'additional extensions to monitor with --watch', list, [])
   .option('--delay', 'wait for async suite definition')
   .option('--allow-uncaught', 'enable uncaught errors to propagate');
```
Comment (dasilvacontin) [inline bin/_mocha]: The `;` at the end of this line is making CI fail.

Comment (dasilvacontin) [inline bin/_mocha]: Sorry, I should have made a commit fixing it directly instead of making a comment. Done. Let's see what CI gives us now.

**Comment by charlierudolph** [comment]
@dasilvacontin @ScottFreeCode all comments addressed

**Comment by boneskull** [comment]
Does this actually tell you which test fails because it's pending, and why?

**Comment by boneskull** [comment]
Needs acceptance by @dasilvacontin

**Comment by dasilvacontin** [comment]
Thanks for the work peeps! 🎈😊

**Comment by ScottFreeCode** [comment]
> Does this actually tell you which test fails because it's pending, and why?

Well I'm kicking myself for not having spotted this before (purely an oversight on my/our part), but I just pulled it down, tried it out and discovered that failure due to `only` or pending tests has no visible difference from passing -- that is, while the result code is affected and this will work in CI, it does not print any indication of the failure, let alone the reason (pending/`only` tests).

I wonder if the point in the code where this is being handled is too late to just insert an `emit('fail', new Error('Pending tests present' / 'Tests marked `.only`'))` and resolve that...

**Comment by dasilvacontin** [comment]
Good insight @ScottFreeCode. What about making the test fail as soon as it is marked as pending?

**Comment by ScottFreeCode** [comment]
Making the test(s) an immediate failure with an error about having been marked `pending` or `only` sounds like it should work.

If that doesn't get implemented in time for the next release, do we want to rebase this into another branch or something so we still have the CLI flag implementation and the tests but aren't publishing a future change of printed output for these flags (instead delaying to publish them till the printed output is ready)? I don't remember whether our printed output is considered a semver-major thing (for features that aren't altogether new like this) or not.

**Comment by dasilvacontin** [comment]
@ScottFreeCode If the printed output only changes when the flag is used, there shouldn't be a problem, right? Since no one was able to consume it before.

**Comment by ScottFreeCode** [comment]
If we don't release `--forbid-*` till they affect printed output, then yeah, the only change occurs when using the new feature and it's semver-minor for sure.

If we release `--forbid-*` initially without effects on printed output and then later add proper printed output to it, so the printed output when using `--forbid-*` changes, is that a backwards compatibility issue or just a bugfix?

**Comment by dasilvacontin** [comment]
Wait. Actually, since when is printed output part of the API? They already have JSON/whatever-reporters for that.

**Comment by ScottFreeCode** [comment]
The JSON reporter has the same issue: it doesn't include any indication about these failures, the return code is the only thing we've got right now.

**Comment by ScottFreeCode** [comment]
> What about making the test fail as soon as it is marked as pending?

Created #2874 for this.

**Pull request merged**